### ISSUE TITLE
Soul Calibur: Projectiles and other hit effects lack color (reported through forum PM)

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#7395]
The reporter gave the following comparison videos:

PSP: https://www.youtube.com/watch?v=kjvOjlfgYJg
PPSSPP (although a pretty old build, but the issue remains): https://www.youtube.com/watch?v=Yz0MIEinRf8&x-yt-ts=1422579428&x-yt-cl=85114404&feature=player_detailpage#t=142

As can be seen, most of the flashy hit effects are more colorful on the real PSP.

Might it be some render-to-CLUT issue?

**Comment by unknownbrackets** [comment]
Can someone with this game on Windows try thy the GE debugger?

-[Unknown]

**Comment by GamerzHell9137** [comment]
This is what you need?
![sou1](https://cloud.githubusercontent.com/assets/2002038/5989209/93cb40ba-a97f-11e4-8460-9a17689ce5dc.png)
![sou2](https://cloud.githubusercontent.com/assets/2002038/5989210/94e02fec-a97f-11e4-9c16-ed29a6a4ea5f.png)
![ss 2015-01-31 at 07 30 35](https://cloud.githubusercontent.com/assets/2002038/5989215/d022482e-a97f-11e4-9309-8b27bda94aea.png)

**Comment by unknownbrackets** [comment]
Yes, but I want to see what the other tabs look like (mainly lighting and texture, but also settings.)

-[Unknown]

**Comment by GamerzHell9137** [comment]
Here

![ss 2015-01-31 at 08 07 26](https://cloud.githubusercontent.com/assets/2002038/5989337/ecb7e6ba-a984-11e4-91ca-de1b4aeca70b.png)

http://puu.sh/fhUCX/ecb77dc96c.txt

**Comment by unknownbrackets** [comment]
Well:
Texture L0 addr	09276cb0, w=128
CLUT	0927acb0

Could be a render-to-clut, but must be followed by a memcpy or something if so.

Hmm:
Tex format	000105

Debugger needs some more formatting there, I guess.

-[Unknown]

**Comment by AizerMortenort** [comment]
@hrydgard  can this be fixed?

**Comment by AizerMortenort** [comment]
Any progress?

**Comment by hrydgard** [comment]
No. If there would be any, you'd see a fix being committed.

**Comment by unknownbrackets** [comment]
Has this improved at all?  I'm wondering if the render-to-clut stuff merged recently will have any impact.

-[Unknown]

**Comment by daniel229** [comment]
Still white color.

**Comment by unknownbrackets** [comment]
The texture itself is all white right?  So it must be colored by lighting?  Unfortunately the link to the other tabs has expired.

-[Unknown]

**Comment by daniel229** [comment]
They are all white.
![6](https://cloud.githubusercontent.com/assets/3481559/12069692/ab2ff00a-b07e-11e5-9781-8e38f689af03.png)

others
https://gist.github.com/daniel229/cea63cec86f0f311403c

**Comment by unknownbrackets** [comment]
I'm seeing again a bit in tex format, dunno what it does or if it does nothing...

This blends the white purely additively on the screen, but I see no color.  I really think the CLUT is wrong.

Does it work in softgpu?

-[Unknown]

**Comment by daniel229** [comment]
No,it doesn't.

**Comment by unknownbrackets** [comment]
Hmm.  Does it work in JPCSP?  JPCSP doesn't seem to recognize that other bit either.  If it doesn't work there, maybe there's some magic CLUT bit that does something crazy.

Another idea: if you get to the GE debugger thing above (they are all white), and then go to texture, and change the CLUT address from whatever it is (e.g. 0926e740) to that + 0x200 (so e.g. 0926e940), does it show anything interesting?

It must be getting the red/whatever color from /somewhere/... maybe it's one of the unknown registers.

-[Unknown]

**Comment by daniel229** [comment]
jpcsp does not work either.

This texture is brown? and it should be displayed as red.
![02](https://cloud.githubusercontent.com/assets/3481559/12073048/1176d766-b142-11e5-9566-5fdd3b400115.png)

This white texture should be displayed as yellow
![4](https://cloud.githubusercontent.com/assets/3481559/12073059/5d042ef4-b142-11e5-8072-f45d777eec5a.png)

+ 0x200 to the CLUT address did not change anything.

**Comment by unknownbrackets** [comment]
Hmm.  Maybe `(texformat >> 8)` masks out certain bits of the texture somehow?

Oh, sorry, you might have to click "Step into" after changing the clut address to update the preview.  Not sure.  If so, should probably fix that...

-[Unknown]

**Comment by daniel229** [comment]
Still not seeing nay change.

That brown texture is just before the white texture,The white txture is what it draw.The brown one I don't know where it draw to.
![01](https://cloud.githubusercontent.com/assets/3481559/12073194/ce766220-b14a-11e5-8c42-f5b7bfe31c41.png)

**Comment by hrydgard** [comment]
The brown and white textures are probably the same texture with different CLUTs. That brown might even look a lot like red when drawn additively?

Yeah, that bit in texfmt is mysterious .. Maybe change the starting point in the CLUT?  I guess that's what you were thinking with the 0x200 offset unknown?

**Comment by unknownbrackets** [comment]
Just wondered if there was colored data somewhere nearby, in case the 0x100 bit in the tex format did something like offset the CLUT?  Seems weird.  Actually, it would be in the loaded CLUT as you say, probably.  Dunno.

-[Unknown]

**Comment by unknownbrackets** [comment]
I did some testing on a PSP and:

 * Still can't figure out what the 0x100 bit in tex format does, if anything.
 * White is the correct result based on the [GE registers from @daniel229](https://gist.github.com/daniel229/cea63cec86f0f311403c#file-soul-calibur-projectiles-and-other-hit-effects-lack-color-reported-through-forum-pm-7395-L23) - simulated the same lighting, texture, framebuffer, etc. settings.
 * If material update is set to 1 (or that bit is set, at least), then red is the correct result; but it's set to 0.
 * Aside from dithering, the test matches PSP vs PPSSPP.

Basically, the MODULATE tex func is being used here to make it red, but the update flag needs to be set to consider the vertex colors.

I even tried setting update to 1, drawing, setting it back to 0 (without resending any colors), and drawing again - in case something tricky happens there.  No dice.

So I think that means one of these things:

1. A bug in PPSSPP is causing the game to use the wrong update value (either never writing/executing the correct one, or incorrectly writing/executing the 0.)

2. There's some tricky patching / sceGeSetCmd / etc. happening while the list is running.

3. Some weird glitch that my test is failing to trigger, and the game just accidentally works (seems unlikely, but who knows.)

4. Something else should be making it red that I've missed.

I suspect the CLUT is not involved here (unless it's the thing I've missed.)

-[Unknown]

**Comment by unknownbrackets** [comment]
Could you try exporting a GE debugger dump on PC?  I'm wondering if there's another draw involved here.

To do this, open the game and select Debug -> GE debugger..., then when it's displaying the scene, press Record in the top right.  After a second or so, it'll finish and save a trace of the drawing activity.

After that, check the memstick/PSP/SYSTEM/DUMP folder and it'll have created a file named something like "ULES12345_0000.ppdmp".  You can zip that and then drag and drop it into a reply here.

-[Unknown]

**Comment by ChurroBoi** [comment]
@unknownbrackets Here you go, never used GE Debugger so hopefully I didn't take any missteps. 

[ULUS10457_0006.zip](https://github.com/hrydgard/ppsspp/files/2087610/ULUS10457_0006.zip)

**Comment by unknownbrackets** [comment]
This is perfect, thanks.  Any reference pictures of what this attack should look like on a real PSP?

Texturing from Texture L0: 0x088e6480 (128x128) in this dump.  It's using a grayscale CLUT, not unexpected.  All 256 colors of the CLUT are part of the grayscale, though.

Hmm...

-[Unknown]

**Comment by ChurroBoi** [comment]
Unfortunately I don't own a real PSP and finding gameplay of this in above average quality is a bit difficult haha. Here is the closest thing I could get, Siegfrieds attacks are blue. Also just to note, when a fighter hits the ground in the original PSP there is a brown cloud of dust, however in PPSSPP it is pure white, I think the issue isn't weapon specific. 
![soulcalibur siegfried](https://user-images.githubusercontent.com/39454123/41199743-c0717338-6cda-11e8-8750-b132b8115363.jpg)

**Comment by unknownbrackets** [comment]
Hmm, this also seems like a good capture of it, more or less:

https://youtu.be/sqJejB3Ez1o?t=20s

It's from 2009, so probably not from an emulator.

Worth noting: since this is blue and the other is red, it definitely can't be some sort of swizzle or mask: the same bit is set for both.  For the same reason, it also can't be dithering: the same dither matrix is used for this blue effect as the red one.

I also didn't see any cases of red being rendered offscreen, e.g. some sort of VRAM/RAM interaction.  But this bug occurs in the software renderer anyway.

My strong suspicion is that the CLUT is somehow wrong (but not sure how - maybe rendered in some previous frame?), because the effect is also too opaque.  Unless the GE registers are wrong, I think the alpha values should not all be a constant 255...

-[Unknown]

**Comment by hrydgard** [comment]
And there's no red / blue in the lighting or in vertex colors? We know we have some lighting bugs, for example in Virtua Tennis the lighting is only correct in the software renderer...

**Comment by unknownbrackets** [comment]
Nope.  But, hmm.  Actually, if I disable lighting entirely it looks bluish, maybe the right color, but too faint.  Changing it to no lighting and using 2.0 * src.a looks like it could even be right, still seems a bit too faint.  It does have color doubling enabled...

Maybe I need to dig up that test and see if I can make sure it's accurate.  Maybe color doubling does something weird like ignoring lighting for the doubling?

-[Unknown]

**Comment by ChurroBoi** [comment]
Have investigated a bit more and it seems nearly ALL graphical effects have been 'white washed'
On the original PSP at the start of a round there is small yellow nova effect when the text fades, and on a KO a large orange nova effect (linked below). All special guard break attacks have a bright lightning effect which matches the colour of the current characters default weapon trail colour and Counter hits flash a light purple, and when characters hit the floor the is a puff of brown dust. 
On PPSSPP however, ALL of these effects are essentially bleached white, the only effect intact is the standard hit effect. Whatever bug is in play seems to affect everything across the board. 
Unsure if helpful but thanks for continuing to look into it. 
![soulcalibur-graphics](https://user-images.githubusercontent.com/39454123/41212438-424938b8-6d81-11e8-8fdd-d381c9a2cd8b.jpg)
![soulcalibur-graphics2](https://user-images.githubusercontent.com/39454123/41212467-825a3c90-6d81-11e8-9435-adc4aad5610a.jpg)

**Comment by hrydgard** [comment]
@unknownbrackets  in https://github.com/hrydgard/ppsspp/issues/7395#issuecomment-251018769 you mention that it looks right if you modify the update flags .  Maybe there's some combination of update flag bits that's not valid on the hardware? or corruption happens of the display list on the command that sets the update flags? (seems exceedingly unlikely though)..

**Comment by unknownbrackets** [comment]
The update flag [doesn't have extra bits](https://github.com/hrydgard/ppsspp/blob/master/Windows/GEDebugger/TabState.cpp#L480-L483), though - that's why I've been so paranoid about logging any extra bits.

Unfortunately, I didn't keep the test so will need to recreate.  But I'm pretty sure I'd accounted for doubling...

In the GE dump, at least within PPSSPP, changing the update flag in the GE debugger after breaking on 0x088e6480 makes it super transparent, but I can't really tell if that's correct or not.  It doesn't look terribly blue, but it's very faint in any case.

-[Unknown]

**Comment by unknownbrackets** [comment]
So, executing the dump as-is on a PSP nets this:

![playback expected](https://user-images.githubusercontent.com/191233/43994224-6845ebc2-9d4e-11e8-8aa8-d44f21e18b66.png)

Which could mean:
 * Recording, or the playback I've just created, is buggy.
 * Emulation is causing the game to behave differently / not write proper GE commands.
 * We're missing a render to CLUT or some other operation, and so it was not included in the recording.
 * This frame is actually supposed to be white, and the slow PSP LCD is in play here (unlikely.)

-[Unknown]

**Comment by hrydgard** [comment]
Huh! Yeah now that's really interesting...

**Comment by unknownbrackets** [comment]
GE dump creation has improved a bit and might catch some of those cases (specifically the CLUT render) better.  Is it possible to create a new GE frame dump of this issue?

That said, since it doesn't work in softgpu I'm not sure what it can be...

-[Unknown]

**Comment by XiaomiRedmi9** [comment]
GE Dump.
[SoulCalibur.zip](https://github.com/hrydgard/ppsspp/files/7070307/SoulCalibur.zip)

**Comment by SangQuangPhung** [comment]
![image](https://github.com/hrydgard/ppsspp/assets/137878765/c1abd0f0-1b20-4e78-be76-764d285a8349)
Today I'm giving a little info on flashy slashes hoping this will improve all characters and custom weapons getting their slash effect color back. In this picture you are seeing Nightmare swinging his sword flashily because he uses the "Soul Edge(Final Form)" equipped from Customize (this sword I don't know how to get it because I'm using the file save unlocks all on gamefaq).
And that's all the information about the "flashy slash" that I discovered hope it will help the Devs to improve the bugs on this topic in the future.

**Comment by unknownbrackets** [comment]
That's very interesting.  Does that mean this bug isn't a bug - it's supposed to be white until you get certain equipment?

Can anyone validate on an actual PSP?

-[Unknown]

**Comment by GamerzHell9137** [comment]
> That's very interesting. Does that mean this bug isn't a bug - it's supposed to be white until you get certain equipment?
> 
> Can anyone validate on an actual PSP?
> 
> -[Unknown]
Can compare on PS Vita (Adrenaline) and can confirm that the effect is not properly emulated on PPSSPP. Still the same things posted above apply. Both PPSSPP and PS Vita use a new save and compared quick battles btw them. On PPSSPP its white while on PS Vita its a transparent red effect.


PPSSPP for comparison
![image](https://github.com/hrydgard/ppsspp/assets/2002038/6bdf7f92-5356-4250-804a-22c300cd7ca7)

**Comment by hrydgard** [comment]
Yeah, seems likely that Nightmare's attack above is just a special case that renders differently.

This one remains one of the weirdest rendering mysteries...

**Comment by hrydgard** [comment]
I know from sources that the PS4/5 remaster has to hack around it, and early videos of it even show that it has the same problem: https://youtu.be/NSQrQmT6bTQ?t=215 

So it's either a game or a hardware bug and if the former, the OS is probably patching the game to fix it.

**Comment by hrydgard** [comment]
Found another video showing what it looks like on hardware: https://www.youtube.com/watch?v=-oO6DUtfimc, I think it's a bit clearer than the previous linked one.

I tried a different quick hack where we simply force update |= 1 throughout the game, doesn't seem to hurt other graphics much, the result has the correct color but is too faint, similar to previous results by unknown, maybe by about half.
